jdaviz/configs/cubeviz/plugins/cube_listener.py

```diff
@@ -0,0 +1,150 @@
+    notes = [["A2"]]
+    score = Score(notes, duration)
+    # set up spectralizer generator
+    generator = Spectralizer(samprate=srate)
+
+    # Lets pick the mapping frequency range for the spectrum...
+    generator.modify_preset({'min_freq': fmin, 'max_freq': fmax,
+                             'fit_spec_multiples': False,
+                             'interpolation_type': 'preserve_power',
+                             'equal_loudness_normalisation': eln})
+
+    data = {'spectrum': [spec], 'pitch': [1]}
+
+    # again, use maximal range for the mapped parameters
+    lims = {'spectrum': ('0', '100')}
+
+    # set up source
+    sources = Events(data.keys())
+    sources.fromdict(data)
+    sources.apply_mapping_functions(map_lims=lims)
+
+    # render and play sonification!
+    soni = Sonification(score, sources, generator, system, samprate=srate)
+    soni.render()
+    soni._make_seamless(overlap)
+
+    return soni.loop_channels['0'].values
+
+
+class CubeListenerData:
+    def __init__(self, cube, wlens, samplerate=44100, duration=1, overlap=0.05, buffsize=1024,
+                 bdepth=16, wl_bounds=None, wl_unit=None, audfrqmin=50, audfrqmax=1500,
+                 eln=False, vol=None):
+        self.siglen = int(samplerate*(duration-overlap))
+        self.cube = cube
+        self.dur = duration
+        self.bdepth = bdepth
+        self.srate = samplerate
+        self.maxval = pow(2, bdepth-1) - 1
+        self.fadedx = 0
+
+        if vol is None:
+            self.atten_level = 1
+        else:
+            self.atten_level = int(np.clip((vol/100)**2, MINVOL, 1))
+
+        self.wl_bounds = wl_bounds
+        self.wl_unit = wl_unit
+        self.wlens = wlens
+
+        # control fades
+        fade = np.linspace(0, 1, buffsize+1)
+        self.ifade = fade[:-1]
+        self.ofade = fade[::-1][:-1]
+
+        # mapping frequency limits in Hz
+        self.audfrqmin = audfrqmin
+        self.audfrqmax = audfrqmax
+
+        # do we normalise for equal loudness?
+        self.eln = eln
+
+        self.idx1 = 0
+        self.idx2 = 0
+        self.cbuff = False
+        self.cursig = np.zeros(self.siglen, dtype='int16')
+        self.newsig = np.zeros(self.siglen, dtype='int16')
+
+        if self.cursig.nbytes * pow(1024, -3) > 2:
+            raise Exception("Cube projected to be > 2Gb!")
+
+        self.sigcube = np.zeros((*self.cube.shape[:2], self.siglen), dtype='int16')
+
+    def set_wl_bounds(self, w1, w2):
+        """
+        set the wavelength bounds for indexing spectra
+        """
+        wsrt = np.sort([w1, w2])
+        self.wl_bounds = tuple(wsrt)
+
+    def audify_cube(self):
+        """
+        Iterate through the cube, convert each spectrum to a signal, and store
+        in class attributes
+        """
+        lo2hi = self.wlens.argsort()[::-1]
+
+        t0 = time.time()
+        for i in tqdm(range(self.cube.shape[0])):
+            for j in range(self.cube.shape[1]):
```

Comment:
Any time I see nested for loops it makes me twitch - I forget how long this takes, but I would advocate for a follow-up ticket to `multiprocess` this (given that it doesn't seem amenable to vectorization) like we do for cube model fitting.

Reply:
Good call, tracking that issue at #3339 .